molten-cache/src/main/java/com/hotels/molten/cache/ReactiveCache.java

```diff
@@ -43,4 +45,46 @@
      * @return a completion once the operation has finished
      */
     Mono<Void> put(K key, V value);
+
+    /**
+     * Return the value stored to {@code key} if any,
+     * otherwise subscribe to the given {@code Mono}
+     * and store its return value in the cache.
+     *
+     * <pre>
+     * return Mono.fromCallable(() -&gt; expensiveServiceCall(key))
+     *   .as(serviceCallCache.cachingWith(key));
+     * </pre>
+     *
+     * @param key the cache key
+     * @return a function
+     */
+    default Function<Mono<V>, Mono<V>> cachingWith(K key) {
+        return cachingWith(key, Function.identity(), Function.identity());
+    }
+
+    /**
+     * Return {@code valueFromConverter.apply(value)}, if {@code value} already stored to {@code key},
+     * otherwise subscribe to the given {@code Mono}
+     * and store its return value as {@code valueToConverter.apply(returnValue)} in the cache.
+     *
+     * <p>Useful for negative cache implementations.</p>
+     *
+     * @param key the cache key
+     * @param valueToConverter transforming the non-cached value to be storable, must not return null
+     * @param valueFromConverter transforming the cached value
+     * @param <U> type of the exposed value
+     * @return a function
+     */
+    default <U> Function<Mono<U>, Mono<U>> cachingWith(K key,
+                                                       Function<U, V> valueToConverter,
+                                                       Function<V, U> valueFromConverter) {
+        return nonCachedMono -> get(key)
+            .switchIfEmpty(nonCachedMono
+                .map(valueToConverter)
+                .switchIfEmpty(Mono.defer(() -> Mono.just(valueToConverter.apply(null))))
```

Comment:
But when `valueToConverter` is `Function.identity()`, it will return null. If we support the possibility of caching the result of an empty Mono, we should also support to keep it empty and not caching it. 
And of course, please add it to the javadocs.

Reply:
I see your point. It is not defined yet if `ReactiveCache` supports null value. Actually it depends on the implementation. However even if we put null value to a non-null key and the implementation allows it, getting the value by its key results an empty `Mono`, which cannot be distinguished from a missing cache hit.

I do agree that an empty `Mono` should still be handled in `cachingWith`, so it now handles if `valueToConverter` returns null. On the other hand the javadoc discourage you to do that.

Do you think it's okay?

Reply:
My main concern is that `cachingWith(K key)` throws NPE if the `nonCachedMono` returns `Mono.empty()`, as `valueToConverter = Function.identity()` will return null, but then it's passed to `Mono.just(null)`, which doesn't handle null. 
For that simple case, I would expect that the "emptiness" is not cached, just like an error.

And I just noticed that you've already fixed it. 😅 
On the javadoc side, it's okay to me 👍